### PR TITLE
[SPARK-31062][K8S][TESTS] Improve spark decommissioning k8s test reliability

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Replace a sleep with waiting for the first collect to happen to try and make the K8s test code more reliable.

### Why are the changes needed?

Currently the Decommissioning test appears to be flaky in Jenkins.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Ran K8s test suite in a loop on minikube on my desktop for 10 iterations without this test failing on any of the runs.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you so much, @holdenk !

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
It passed once. I re-triggerred to see the stability.

**Comment by dongjoon-hyun** [comment]
The failure is irrelevant to this PR.
```
KubernetesSuite:
- Run SparkPi with no resources *** FAILED ***
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by holdenk** [comment]
I noticed that test flaking some too, I'll try and tackle it as well but I figured I'd do this one first.
Jenkins retest this please.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Are log files for executors available on artifact of Jenkins build? Without the information it doesn't seem to be easy to look into the issue with details.

**Comment by dongjoon-hyun** [comment]
For now, it looks better. l take a look tomorrow and trigger a little more. Thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala [hunk 1]
```diff
@@ -32,8 +32,9 @@ private[spark] trait DecommissionSuite { k8sSuite: KubernetesSuite =>
     runSparkApplicationAndVerifyCompletion(
       appResource = PYSPARK_DECOMISSIONING,
       mainClass = "",
-      expectedLogOnCompletion = Seq("decommissioning executor",
-        "Finished waiting, stopping Spark"),
+      expectedLogOnCompletion = Seq(
+        "Finished waiting, stopping Spark",
+        "decommissioning executor"),
```
Comment (ScrapCodes) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala]: Hi Holden, in one of the failed runs on Jenkins, I have found this text did not appear at all in the Logs. In order to fix the flakiness, the cause might be deeper than the order in which they appear.

Comment (ScrapCodes) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DecommissionSuite.scala]: Thanks for looking at this issue though !

**Comment by dongjoon-hyun** [comment]
Unfortunately, `Jenkins` is still complaining.
```
- Test basic decommissioning *** FAILED ***
  The code passed to eventually never returned normally. Attempted 126 times over 2.0171248232833334 minutes. Last failure message: "++ id -u
```

**Comment by holdenk** [comment]
So I think this PR does fix a race condition in the test, but I think there is a bug in how we determine if a pod is deleted which has a race condition as well. If folks are OK with it I'd like to do this fix now based on it not flaking on my local env and I'll follow up with a fix around the snapshot/deleted condition race.

**Comment by holdenk** [comment]
Merged to master so we can have a more reliable test for now. I'll follow up with the race condition on pod deletion in another PR ASAP.

**Pull request closed without merging**